### ISSUE TITLE
Memory usage estimates are overly optimistic thanks to object overhead

### ISSUE THREAD

**Issue by sesuncedu** [RoaringBitmap/RoaringBitmap#257]
RoaringBitmap provides methods to calculate the estimated memory usage of a particular set of integers.   The calculated value does not include per-object overhead, which causes the estimates to be negatively biased.  The error can be quite large, and has practical implications - for example, when trying to use a tiling approach when comparing a large number of bitmaps.   

I have implemented some experimental code to calculate precise values for memory usage (the only layout I am currently handling is hotspot 64 bit  with 32-bit compressed oops. 

The test data I used is from a project I'm currently working on to analyze the contents of maven central. There were 865,727 sets (after eliminating exact duplicates), with a total cardinality of 476,339,914;  each bitmap represents a set of integers that identify the name&hash of the classes in a particular jar.  These identifiers were not assigned in a particularly roaring oriented fashion.  

Once precise and estimated sizes for the bitmaps were obtained, new bitmaps were created by mapping the original identifiers to a new range of sequentially allocated values, which were then run optimized and packed. 

For the untransformed dataset,  estimated size was almost entirely explained by the cardinality (R^2 = 0.97). Actual size was much less correlated (R^2 = 0.60). 

For the untransformed identifiers, some descriptive statistics:
```
summary(roar)
  cardinality          estimate         precise      
 Min.   :     1.0   Min.   :    16   Min.   :   136  
 1st Qu.:    12.0   1st Qu.:   104   1st Qu.:   720  
 Median :    45.0   Median :   360   Median :  2448  
 Mean   :   550.2   Mean   :  1955   Mean   :  8334  
 3rd Qu.:   164.0   3rd Qu.:  1236   3rd Qu.:  8176  
 Max.   :117984.0   Max.   :241316   Max.   :279696  

summary(roar$precise/roar$estimate)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  1.159   6.608   6.775   6.662   6.944   8.500 
quantile(roar$precise/roar$estimate,seq(0,1,0.1))
      0%      10%      20%      30%      40%      50%      60%      70%      80%      90%     100% 
1.159045 6.140872 6.540468 6.655022 6.721649 6.775120 6.830769 6.900000 7.000000 7.333333 8.500000 
```
For the transformed identifiers:
```
summary(roar2)
  cardinality          estimate           precise        
 Min.   :     1.0   Min.   :    16.0   Min.   :   136.0  
 1st Qu.:    12.0   1st Qu.:    26.0   1st Qu.:   184.0  
 Median :    45.0   Median :    58.0   Median :   256.0  
 Mean   :   550.2   Mean   :   712.4   Mean   :   966.4  
 3rd Qu.:   164.0   3rd Qu.:   180.0   3rd Qu.:   464.0  
 Max.   :117984.0   Max.   :147352.0   Max.   :151432.0  

summary(roar2$precise/roar2$estimate)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  1.010   2.454   4.545   4.577   6.800   8.500 

quantile(roar2$precise/roar2$estimate,seq(0,1,0.1))
      0%      10%      20%      30%      40%      50%      60%      70%      80%      90%     100% 
1.009515 1.450000 2.082840 2.846154 3.644444 4.545455 5.391304 6.181818 7.555556 7.555556 8.500000 
```
For the transformed case,  less than 0.001% of the rations were below 1.01. 

Example (intArray size is the size of a hypothetical object wrapping an int[])
```
19:14:59:107 [main] INFO ShowBitmapSizes - false win
19:14:59:107 [main] INFO ShowBitmapSizes -              original: est.  10,126, precise  46,568, intArray  10,200, card =   2,542, ebpe = 31.868, pbpe = 146.555
19:14:59:107 [main] INFO ShowBitmapSizes - false win
19:14:59:107 [main] INFO ShowBitmapSizes -    original,runs,trim: est.  10,126, precise  46,568, intArray  10,200, card =   2,542, ebpe = 31.868, pbpe = 146.555
19:14:59:108 [main] INFO ShowBitmapSizes - false win
19:14:59:108 [main] INFO ShowBitmapSizes -                 renum: est.   5,524, precise  10,320, intArray  10,200, card =   2,542, ebpe = 17.385, pbpe = 32.478
19:14:59:108 [main] INFO ShowBitmapSizes - win
19:14:59:108 [main] INFO ShowBitmapSizes -       renum,runs,trim: est.   4,656, precise   7,816, intArray  10,200, card =   2,542, ebpe = 14.653, pbpe = 24.598
...
19:14:59:111 [main] INFO ShowBitmapSizes - win
19:14:59:111 [main] INFO ShowBitmapSizes -                 renum: est.   1,264, precise   1,600, intArray   2,536, card =     625, ebpe = 16.179, pbpe = 20.480
19:14:59:111 [main] INFO ShowBitmapSizes - win
19:14:59:111 [main] INFO ShowBitmapSizes -       renum,runs,trim: est.   1,264, precise   1,384, intArray   2,536, card =     625, ebpe = 16.179, pbpe = 17.715

19:14:59:116 [main] INFO ShowBitmapSizes - false win
19:14:59:116 [main] INFO ShowBitmapSizes -              original: est.  14,038, precise  52,120, intArray  17,488, card =   4,363, ebpe = 25.740, pbpe = 95.567
19:14:59:116 [main] INFO ShowBitmapSizes - false win
19:14:59:116 [main] INFO ShowBitmapSizes -    original,runs,trim: est.  14,038, precise  52,120, intArray  17,488, card =   4,363, ebpe = 25.740, pbpe = 95.567
19:14:59:116 [main] INFO ShowBitmapSizes - win
19:14:59:116 [main] INFO ShowBitmapSizes -                 renum: est.   8,908, precise  11,664, intArray  17,488, card =   4,363, ebpe = 16.334, pbpe = 21.387
19:14:59:116 [main] INFO ShowBitmapSizes - win
19:14:59:116 [main] INFO ShowBitmapSizes -       renum,runs,trim: est.   8,890, precise  10,224, intArray  17,488, card =   4,363, ebpe = 16.301, pbpe = 18.747

```

**Comment by lemire** [comment]
You do not provide your dataset, but from your description, I understand that your sets of made of random integers. That's an adversarial use case. That is, you should not be using a bitmap/bitset in such a scenario.

I don't think it was explicitly stated. However, the README.md files does indicate that a bitset is not always the right data structure. I have elaborated in the project documentation, but you may want to check out my talk on this subject...

https://youtu.be/ubykHUyNi_0

In general, hash values are not compressible. 

I'm closing this. Reopen if my latest commit does not answer your query.

**Comment by sesuncedu** [comment]
TL;DR:  As discussed above, this experiment was run both with the original identifiers, and with remapped identifiers, with the remapped identifiers sequentially assigned in a roar-friendly fashion.  Remapped identifiers produced much better local density; however although the difference between estimate and actual memory  was lower than with un-remapped identifiers, it remains high.           

To clarify:  the contents of the bitmaps in both versions of the data are sequentially assigned integers.  

In the first case, the identifiers were assigned in an order not corresponding to set containment, which, as you note, is unlikely to yield sets with sufficiently local density to take advantage of RoaringBitmap, and as expected did not give especially good results.
However, even in this case the difference between estimated and precise sizes was notable. 

With a total of 476,339,914 elements in 865,727 sets,  the estimated size predicted a mean number of bits per element of ~28.4.  However, using the precise size gives a mean value of ~121.2 bits per element. The mean estimated size was 1,955 bytes; the mean precise size was 8,334 bytes.

The estimated sizes predict that ~3.9% of all sets use less than 32bits per value.  Using the precise sizes show that only ~0.46% of all sets used 32 bits or less.

Since the initial identifier assignment was expected to give poor results, a renumbering process was used that mapped the original identifiers to sequentially allocated  values, on a set by set basis.  This was expected to give much better results, ( given that many sets are proper subsets of, or to have substantial overlap with, other sets).  The predicted improvement was in fact observed. However, the differences between estimated and precise sizes remained substantial. 

The mean  estimated bpe dropped to 10.4 - however the mean precise bpe was 14.0512. Mean estimated size dropped to 712 bytes, with mean precise size 966.

~37% of all sets had a precise bpe of <32, compared to an estimate of ~83%

Ignoring the smallest and largest 2.5% of sets, the mean difference was 228 bytes, compared to a mean estimated size of 271 bytes.  The difference is only weakly explained by cardinality (R^2 = 0.33), making the estimate problematic for use e.g. when using a tiling approach for cache optimizations.

**Comment by lemire** [comment]
@sesuncedu 

The library provides an estimate with the following detailed instructions...


>  Estimate of the memory usage of this data structure. This can be expected to be within 1% of the true memory usage in common usage scenarios. If exact measures are needed, we recommend using dedicated libraries such as ehcache-sizeofengine.  In adversarial cases, this estimate may be 10x the actual memory usage. (...)   Note that you can serialize your Roaring Bitmaps to disk and then construct ImmutableRoaringBitmap instances from a ByteBuffer. In such cases, the Java heap usage will be significantly less than what is reported.

We do not promise to report the exact memory usage. We provide an estimate. We even warn that it might be off by a factor of 10 in problematic cases.

We validate that our estimate is good using several typical dataset...

https://github.com/RoaringBitmap/RoaringBitmap/blob/master/memory/src/test/java/org/roaringbitmap/RealDataDetailedBenchmark.java


That's a lot better than most Java data structures out there... they *never* offer any kind of measure as to how much memory they might be using. We are doing a great job on the use cases that have been submitted to us.

Now... in your case, the estimate is not sufficiently precise for your purposes. Fine. But it is super easy to get a more exact value if you like...

```
import org.ehcache.sizeof.SizeOf;
import org.ehcache.sizeof.impl.AgentSizeOf;

private static final SizeOf sizeOf = AgentSizeOf.newInstance(false, true);

int memory_usage = sizeOf.deepSizeOf(mybitmap);
```

Maybe this is inconvenient and you'd like the RoaringBitmap to provide a more accurate estimate... if you'd like to help us improve our estimate, we would be happy to take a pull request. 

But as things stand now, there is no error in the library. An estimate can be off... that's what it is... an estimate. It is not a precise measure.

**Comment by sesuncedu** [comment]
Calculating precise sizes does not involve much more work at run time than the estimated version (the primary changes are adding constant factors, with a few adds and masks to account for alignment padding, many of which can be ignored).  

For example:  The size of a RunContainer is currently calculated as: 

```java
  public int getSizeInBytes() {
    return this.nbrruns * 4 + 4;
  }
```
A precise estimate for hotspot-64bit-compressed-oops would be
```java
 public int getSizeInBytes() {
   long size = 0;
   size += 12; // header size
   size += 4; // size of valueslength field
   size += 4; // size of nbrruns field
   size += 4; //  inter-object padding to 8 byte boundary
   size += 16; // size of valuelength array header
   size += nbrruns.length * 2; // size of valuelength array content (shorts)
   size += 7;   // pad to 8 byte 
   size &= ~7; // boundary
   return size;
  }
```
which reduces to 
```
 public int getSizeInBytes() {
   int size = 47 +  valueslength.length * 2; 
   return size & ~7;
  }
```
The extra work is a fetch of the length field from the array header, and a mask to clear the lower 3 bits.

**Comment by lemire** [comment]
> Calculating precise sizes does not involve much more work at run time

Please see the end of my reply...

`Maybe this is inconvenient and you'd like the RoaringBitmap to provide a more accurate estimate... if you'd like to help us improve our estimate, we would be happy to take a pull request.`

Of course, you'll need to specify that this is the memory usage for a particular JVM because, as far as I know, there is no language-level specification of the memory usage.

So maybe you need a function called

``getSizeInBytesHotspot64bitCompressedOOPS()``

... because if you claim that the returned value  is precise memory usage, you need to be precise as to what the assumed JVM is.

**Comment by sesuncedu** [comment]
```java
getSizeInBytesHotspot64bitCompressedOOPS()
```
In the prototype code I wrote, I am just passing in a helper object to encapsulate the details.   
Hotspot & IBM are easy enough to  handle, which covers most server platforms.   
Using the ehCache sizeof library is a good idea as an option for handling other platforms.

**Comment by lemire** [comment]
> In the prototype code I wrote, I am just passing in a helper object to encapsulate the details.

Something of the sort might be good.

> Using the ehCache sizeof library is a good idea as an option for handling other platforms.

The JVM must be started with a special flag so that the instrumentation can be done. The memory subdirectory has special tests with this instrumentation.

Of course, we will want to validate an eventual PR against ehCache. (We have tests in the memory subdirectory.)

**Comment by sesuncedu** [comment]
It's possible to dynamically attach with openjdk/oracle/IBM via
com.sun.tools.attach.VirtualMachine
though it may require creating a new ClassLoader  and adding tools.jar to
it.

Not as scary as JOLs serviceability agent code, which may start a need to
start a child process using sudo.

Validating against existing test suite makes things  easy. Or hard :-)

**Comment by lemire** [comment]
Thanks @sesuncedu : PR invited.

**Comment by sesuncedu** [comment]
WILCO. I will create a separate issue to cover a some api & implementation choices as I clean things up for a PR.
